src/scenarios_hpc_azure/azure_utils.py

```diff
@@ -474,6 +478,92 @@ def launch_postprocess(
 
         return postprocess_task_ids
 
+    def monitor_and_download(
+        self,
+        timeout_mins: int,
+        dest: str,
+        targets: Optional[list[str]] = None,
+    ) -> list[str]:
+        """Monitors a currently launched job, downloads job's outputs
+        upon completion to `dest/` directory on callers machine. Overwrites
+        name collisions if same jobid is run twice.
+        Can take `targets` strs which, if found in the path of jobs output files,
+        includes it in downloads. `targets` can be filenames or directories.
+
+        Parameters
+        ----------
+        timeout_mins : int
+            how long to monitor the job before raising an error, does NOT
+            halt execution on Azure Batch if this limit is reached, simply
+            stops monitoring and attempts to download what is available.
+        dest : str
+            source destination for downloaded files on users machine.
+            `experiment_name/jobid` directories appended to avoid collisions.
+        targets : list[str], optional
+            list of files/directories from the blob you wish you retrieve,
+            None retrieves the entire experiment/jobid/ directory,
+            containing all output files. by default None
+
+        Returns
+        -------
+        list[str]
+            paths to downloaded files in `dest`
+
+        Raises
+        -------
+        RuntimeError if `timeout_mins` is reached before job completes
+        """
+        # monitor the job, execution passed back when job completes, or after timeout_mins minutes
+        self.azure_client.monitor_job(self.job_id, timeout=timeout_mins)
+        # execution has returned, lets try to download the requested files now
+        output_dir = os.path.join(self._experiment_name, self.job_id)
+        print(
+            (
+                f"{bcolors.WARNING} downloading files from blob {output_dir} "
+                f"to {dest} overwriting any existing files with the same filenames. {bcolors.ENDC}"
+            )
+        )
+        if targets is not None:
+            # get all the output files from the job that just finished
+            all_job_outputs = (
+                self.azure_client.out_cont_client.list_blob_names(
+                    name_starts_with=output_dir
+                )
+            )
+            # then scan to see if `target` is inside each blob path
+            target_blob_paths = [
+                blob_path
+                for blob_path in all_job_outputs
+                if any(target in blob_path for target in targets)
+            ]
+            # cant call download_directory and pass individual file paths
+            # so we must download each file 1 by 1. This is what cfa_azure
+            # does in the backend anyways
+            written_dirs = []
+            for blob in target_blob_paths:
+                if "." in blob:
```

Comment:
hmmmm, why this conditional? 🤔

Reply:
we are using the `download_file()` directory to download specific files that the user specifies in `--targets`. `cfa_azure` has two ways to download something, `download_file()` and `download_directory()`. You can not pass a file path to `download_directory()` and vice versa. Therefore when a user specifies `--targets` we instead download each file individually. The conditional is to check if the path given in `target_blob_paths` is a file, aka `file.txt`, or `file.py`

Reply:
Could this be done using, say `os.path.isfile(blob)`? Current logic is sound enough with some potential edge case like if there's a directory that's called `xyz2.2` then it can potentially break. if `.isfile` doesn't work then it's fine to proceed with it.

Reply:
`os.isfile()` checks your local directories for the file, these files do not exist on the users local machine at the time that line of code is executed, they are downloaded shortly after.